drivers: watchdog: Add watchdog driver support on Renesas RX130

quytranpzz opened the pull request:
Add watchdog driver support on RSK-RX130@512kb (Independent Watchdog Timer).

Users can the counter by 2 modes (select by WDT_RENESAS_RX_IWDT_REGISTER_START_MODE or WDT_RENESAS_RX_IWDT_AUTO_START_MODE)
- Auto-start mode: controlled by option function select register 0 (OFS0).
- Register start mode: controlled by the IWDT registers

sangtranjc committed:
drivers: watchdog: Support Renesas RX independent watchdog timer driver

Add initial support for independent watchdog driver for Renesas RX
with r_iwdt_rx RDP HAL

Signed-off-by: Sang Tran <sang.tran.jc@renesas.com>

sangtranjc committed:
arch: rx: Add NMI vector table for Renesas RX MCU

Add support for non-maskable interrupt (NMI)  vector table for
Renesas RX architecture

Signed-off-by: Sang Tran <sang.tran.jc@renesas.com>

quytranpzz committed:
soc: renesas: rx: enable option function select register 0

Enables OSF0 register select for IWDT driver setting on start mode

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
dts: renesas: rx: add iwdt property node for watchdog driver

Add iwdt property node on dts for watchdog driver

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
boards: renesas: Enable watchdog driver on RSK-RX130-512kb

Enable iwdt dts nodes and iwdtsclk clock on RSK-RX130-512kb

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
tests: drivers: watchdog: use Kconfig options for setup flags and timeout

Replace hard-coded platform definrs with Kconfig marcos

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

duynguyenxa commented on drivers/watchdog/Kconfig.renesas_rx:
```diff
@@ -0,0 +1,104 @@
+
+if IWDT_RENESAS_RX
+
+config IWDT_RX_NMI
+	bool "Non-maskable interrupt for IWDT"
+	default y
+	help
+	  Enable NMI for IWDT.
```
```suggestion
config WDT_RENESAS_RX_IWDT_USE_NMI
	bool "Non-maskable interrupt for IWDT"
	default y
	help
	  Enable NMI for IWDT.
```

duynguyenxa commented on drivers/watchdog/Kconfig.renesas_rx:
```diff
@@ -0,0 +1,104 @@
+# Renesas RX Independent Watchdog (IWDT) configuration
```
```suggestion
# Renesas RX Watchdog configuration
```
This file later maybe will be used for other WDT driver in Renesas RX as well

duynguyenxa commented on soc/renesas/rx/rx130/ofsm.c:
```diff
@@ -15,6 +15,8 @@
  */
 
 #define __OFS_MDE __attribute__((section(".ofs_mde")))
+#define __OFS0    __attribute__((section(".ofs0")))
```
@quytranpzz , although this file is put here but it hasn't been added in the CMakeLists.txt, please help to recheck this

duynguyenxa commented on drivers/watchdog/wdt_iwdt_renesas_rx.c:
```diff
@@ -0,0 +1,224 @@
+	}
+}
+
+static void iwdt_enable_nmi(void)
+{
+	ICU.NMIER.BIT.IWDTEN = 0x1;
+}
```
@quytranpzz , This should be an API of nmi handler in arch layer

duynguyenxa commented on drivers/watchdog/Kconfig.renesas_rx:
```diff
@@ -0,0 +1,104 @@
+	help
+	  Enable NMI for IWDT.
+
+choice
+	prompt "IWDT Start Mode"
+	default IWDT_RENESAS_RX_IWDT
```
@quytranpzz , To be honest, I don't like this KConfig here, for the auto start mode, I think it should be an option under SOC layer.
I think the design should be IWDT mode select, and the KConfig related to it should be under the RX SOC, when the register start mode is chosen, then the driver is enable.
This config option and the option related to the AUTO_START_MODE should be placed under SoC
Else the IWDT driver should only support register start mode

quytranpzz committed:
soc: renesas: rx: Update OFS value in vects.c using Kconfig

OFS values setting for RXv1/RXv2 will be defined in SOC
Kconfig and set in vects.c file

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
include: arch: rx: Move _ebss before noinit to preserve after reset

This change ensure .noinit is excluded from zeroed memory region after
reset

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

thenguyenyf commented on soc/renesas/rx/rx130/Kconfig:
```diff
@@ -9,3 +9,168 @@ config SOC_SERIES_RX130
 	select SOC_EARLY_INIT_HOOK
+
+if SOC_SERIES_RX130
+if WDT_RENESAS_RX_IWDT_AUTO_START_MODE
+
+choice
+	prompt "IWDT OFS0 Start Mode Select"
+	default WDT_RENESAS_RX_STARTMODE_HALT
+
+config WDT_RENESAS_RX_STARTMODE_AUTO_ACTIVATE
+	bool "IWDT is automatically activated in auto-start mode after a reset"
```
Should it placed in `drivers/watchdog/Kconfig.renesas_rx`?

quytranpzz replied:
Hmm. I think it makes sense to place it here since it belongs to soc config value

cfriedt commented on tests/drivers/watchdog/wdt_basic_reset_none/src/main.c:
```diff
@@ -27,6 +27,10 @@
 #define WDT_SETUP_FLAGS		0
 #define WDT_TIMEOUT		K_TICKS(WDT_TIMEOUT_VALUE)
 #define SLEEP_TIME		K_TICKS(CONFIG_TEST_WDT_SLEEP_TIME)
+#elif defined(CONFIG_WDT_RENESAS_RX_IWDT)
+#define WDT_SETUP_FLAGS		0
+#define WDT_TIMEOUT		K_MSEC(WDT_TIMEOUT_VALUE)
+#define SLEEP_TIME		K_MSEC(CONFIG_TEST_WDT_SLEEP_TIME)
```
Hard-coding platform-specific defines inside of generic tests should really be avoided if at all possible.

Ideally for this kind of test, there would be local Kconfig options that are overridden on a per board basis, in `boards/<board>.conf`.

quytranpzz replied:
I have updated the source using Kconfig options for setup flag. Could you please take a look?

cfriedt reviewed (dismissed):
Can this (or possibly another) PR adjust the `samples/drivers/watchdog` to support local Kconfig options instead of hard-coding platform specific defines inside of the source file?

JarmouniA reviewed (dismissed):
Please avoid mixing formatting changes with functional ones in the same commit, and avoid formatting changes completely if they are not required for CI.

quytranpzz commented on tests/drivers/watchdog/wdt_basic_reset_none/src/main.c:
```diff
@@ -80,7 +77,7 @@ static int test_wdt_callback_reset_none(void)
 	wdt_interrupted_flag = 0;
 	for (int i = 0; i < WDT_FEED_TRIES; ++i) {
-		TC_PRINT("Feeding %d\n", i+1);
+		TC_PRINT("Feeding %d\n", i + 1);
```
Thanks, I removed the change

quytranpzz commented on tests/drivers/watchdog/wdt_basic_reset_none/src/main.c:
```diff
@@ -132,7 +128,6 @@ static int test_wdt_bad_window_max(void)
 	return TC_FAIL;
 }
 
-
```
removed change

quytranpzz commented on tests/drivers/watchdog/wdt_basic_reset_none/src/main.c:
```diff
@@ -41,9 +39,8 @@ static volatile int wdt_feed_flag;
-	zassert_equal(WDT_FEED_TRIES, wdt_feed_flag,
-			"%d: Invalid number of feeding (expected: %d)",
-			wdt_feed_flag, WDT_FEED_TRIES);
+	zassert_equal(WDT_FEED_TRIES, wdt_feed_flag, "%d: Invalid number of feeding (expected: %d)",
```
removed change
